### ISSUE TITLE
fft with 1.16.6 is faster than with 1.18.0

### ISSUE THREAD

**Issue by Intro1997** [numpy/numpy#17130]
### Reproducing code example:

```python
import random
import time
# import matplotlib
import numpy as np 
from scipy.fftpack import fft
# import cv2

coordinate = np.ndarray((1920, 1080, 2))
print(coordinate.shape)

for i in range(1920):
    for j in range(1080):
      coordinate[i][j] = (random.randint(0, 1920), random.randint(0, 1080))

print(coordinate[0][0])
# test = ([2.+4.j, 3.+3.j, 4.+4.j, 5.+3.j, 4.+2.j, 3.+1.j, 2.+2.j, 1.+3.j])
start_time = time.time();
res = np.fft.fft(coordinate)
end_time = time.time();
print("spend time: ", (end_time - start_time));
```

### Error message:

The result is :
```
% python2 ./test.py
(1920, 1080, 2)
[ 1022.   435.]
('spend time: ', 0.06316184997558594)
% python3 ./test.py
(1920, 1080, 2)
[1143.  678.]
spend time:  0.18218493461608887
```

### Numpy/Python version information:
The pythons' numpy version are both 1.19.1
Python2: 2.7.16
Python3: 3.7.3

macOS: 10.15.6

**Comment by Intro1997** [comment]
> You cannot "convince" one python version to work with another version's numpy. You need to install for each version.
> 
> If you are using NumPy 1.8.0rc1 in python2, that is a version from Sept 2013. Since then much has changed in NumPy. The last version to support python2 was NumPy 1.16.6, but you will need pip or homebrew to install it. The change could have come from many sources, here are a few:
> 
> * maybe your NumPy is using mkl underneath to provide an optimized fft. You can get a hint by looking at the output of `numpy.show_config()`.
> * somehow the `dtype` for `coordinate` is different, so you are doing a different calculation. You can check this by examining `coordinate.dtype`
> * the code for fft has changed, perhaps for your use case something slowed down

After I changed my python2 and python3's numpy version to 1.16.6, their speed is not much different:
```
intro@~ % python3 
Python 3.8.5 (default, Jul 21 2020, 10:48:26) 
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.version.version
'1.16.6'
>>> exit()
intro@~ % python2 

WARNING: Python 2.7 is not recommended. 
This version is included in macOS for compatibility with legacy software. 
Future versions of macOS will not include Python 2.7. 
Instead, it is recommended that you transition to using 'python3' from within Terminal.

Python 2.7.16 (default, Jul  5 2020, 02:24:03) 
[GCC 4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.29.21) (-macos10.15-objc- on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.version.version
'1.16.6'
>>> exit()
intro@~ % python2 test.py && python3 test.py 
(1920, 1080, 2)
[815  65]
('spend time: ', 0.06374310970306396)
(1920, 1080, 2)
[815  65]
spend time:  0.06189305067062378
```
But when I change the numpy version of python3 from 1.16.6 to 1.18.0, python3 will be slow.
```
intro@~ % python3
Python 3.8.5 (default, Jul 21 2020, 10:48:26) 
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.version.version
'1.18.0'
>>> exit()
intro@~ % python2 test.py && python3 test.py  
(1920, 1080, 2)
[815  65]
('spend time: ', 0.061868069171905515)
(1920, 1080, 2)
[815  65]
spend time:  0.1591855502128601
```
I check the mkl on python2 numpy-1.16.6 and it's not avaliable:
```
intro@~ % python2

WARNING: Python 2.7 is not recommended. 
This version is included in macOS for compatibility with legacy software. 
Future versions of macOS will not include Python 2.7. 
Instead, it is recommended that you transition to using 'python3' from within Terminal.

Python 2.7.16 (default, Jul  5 2020, 02:24:03) 
[GCC 4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.29.21) (-macos10.15-objc- on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.show_config()
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
blis_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
lapack_mkl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
>>> 
```
same as python3:
```
intro@~ % python3     
Python 3.8.5 (default, Jul 21 2020, 10:48:26) 
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.version.version
'1.16.6'
>>> np.show_config()
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
accelerate_info:
    extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
blas_opt_info:
    extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_opt_info:
    extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
>>> 
```
Maybe there are something problem with numpy-1.18.0?

**Comment by mattip** [comment]
I have changed the issue title since we now see that the python version is not involved: NumPy 1.16.6 is faster than NumPy 1.18.0. When you install 1.18.0, is is using the PyPI provided wheel or building from source? There are no wheels available for python 3.8 on versions prior to 1.18.0, so perhaps the issue is that a local build is faster than the downloaded wheel on macOS.

**Comment by Intro1997** [comment]
> I have changed the issue title since we now see that the python version is not involved: NumPy 1.16.6 is faster than NumPy 1.18.0. When you install 1.18.0, is is using the PyPI provided wheel or building from source? There are no wheels available for python 3.8 on versions prior to 1.18.0, so perhaps the issue is that a local build is faster than the downloaded wheel on macOS.

I use the ```python3 -m pip install numpy==1.18.0``` to install the numpy 1.18.0 and 1.16.6. 
As for 1.16.6, the setup info is:
```
Processing ./Library/Caches/pip/wheels/22/6f/a6/069db0c95f9bb6b73e07da014891c58460fef1eb84f49576f0/numpy-1.16.6-cp38-cp38-macosx_10_15_x86_64.whl
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.18.0
    Uninstalling numpy-1.18.0:
      Successfully uninstalled numpy-1.18.0
Successfully installed numpy-1.16.6
```
Could I tell from this information that my numpy-1.16.6 is a locally built version?

**Comment by seberg** [comment]
In 1.17, we got pocketfft for much faster FFTs, so that is probably the reason.

I profiled NumPy master, and it is not super obvious to me this is a large issue.  The example calls millions of very short FFTs (length 2), so the overhead of each of those FFT must be larger.  I.e. most time is spend in the `pass_all` function: https://github.com/numpy/numpy/blob/ec62be21cf109dac436f9a68a84fbd674c1faebb/numpy/fft/_pocketfft.c#L883

There are some allocations and probably other things which do not have to be repeated every time, moving them out could probably speed things up by 20+% at least (it is also using `malloc`).  Also just choosing the plan every time in `cfft_forward` might take a decent chunk of the total runtime.

In any case, I am unsure what the tradeoffs are. I do not think we have to try and get length 2 FFTs to the same level they were.  But making them slightly faster could be nice, and I do not know whether the difference is gone for 124 FFTs and we may care about those.

@mreineck do you have an opinion on this?

@Intro1997 maybe one thing if you are interested to getting fast FFTs in that use-case.  You might try FFTW, it is fast and I somewhat think it may be fast specifically in this type of use case (huge number of identical FFTs).

**Comment by mreineck** [comment]
> @mreineck do you have an opinion on this?

If I understand the test case correctly, it is executing only length-2 FFTs, which is a pathological case IMO.
In case that this is really what was intended, the best solution is probably not to go via FFTW, but just to compute
`coordinate[:,:,0] + coordinate[:,:,1]`
and
`coordinate[:,:,0] - coordinate[:,:,1]`
which is the desired result :-)

But my guess/hope is that there was a misunderstanding about what `numpy.fft.fft` does... I myself thought for a long time that it would actually transform all axes, not just the last one.

Optimizing this kind of invocation may be possible, but I expect it would have to be done in the Python wrapper already, not in the C library itself, in order to have full effect. Personally I don't think it's worth the trouble.

**Comment by seberg** [comment]
Yeah, I think the main difference might be that the `pass_all` function is optimal for long(er) FFTs and just a bit heavy weight in such a pathological case.  There is probably room for some simple speed-ups, but I am not holding my breath that they will make a big difference.  (E.g. we could use python allocation, to be faster for very short FFTs)

**Comment by Intro1997** [comment]
> In 1.17, we got pocketfft for much faster FFTs, so that is probably the reason.
> 
> I profiled NumPy master, and it is not super obvious to me this is a large issue. The example calls millions of very short FFTs (length 2), so the overhead of each of those FFT must be larger. I.e. most time is spend in the `pass_all` function:
> 
> https://github.com/numpy/numpy/blob/ec62be21cf109dac436f9a68a84fbd674c1faebb/numpy/fft/_pocketfft.c#L883
> 
> There are some allocations and probably other things which do not have to be repeated every time, moving them out could probably speed things up by 20+% at least (it is also using `malloc`). Also just choosing the plan every time in `cfft_forward` might take a decent chunk of the total runtime.
> 
> In any case, I am unsure what the tradeoffs are. I do not think we have to try and get length 2 FFTs to the same level they were. But making them slightly faster could be nice, and I do not know whether the difference is gone for 124 FFTs and we may care about those.
> 
> @mreineck do you have an opinion on this?
> 
> @Intro1997 maybe one thing if you are interested to getting fast FFTs in that use-case. You might try FFTW, it is fast and I somewhat think it may be fast specifically in this type of use case (huge number of identical FFTs).

In fact, this issue was founded when I compared FFTW and np.fft.fft >.<

**Comment by mreineck** [comment]
OK, but did you really originally want to benchmark length-2 FFTs? I don't understand the potential application for this ...

**Comment by mreineck** [comment]
Since `pocketfft` has a more complex internal decision tree (e.g. so it can use Bluestein's algorithm for near-prime sizes), it also requires more complex memory management and control flow. This is of course complete overkill for transform of lengths below, say, 20. Having persistent plans (like FFTW does) can help a lot here, but also adds much complexity.

That said, there are a few low-hanging fruit: the `malloc/free` in `pass_all`, `rfftp_forward` and `rfftp_backward` could be avoided for small sizes by using a fixed-size local array. But skipping the memory allocation for the twiddle factors for tiny sizes is already way more complicated.

**Comment by mehrdadn** [comment]
Hi there! I recently noticed a dramatic drop in FFT and convolution performance when using NumPy, and after learning about the migration to pocketfft, realized it's the culprit. Since the above example appears to use small arrays, I thought I'd share another benchmark that illustrates this with larger arrays. The benchmark is below:
```
import sys
from timeit import default_timer
import numpy.fft

fpl = getattr(numpy.fft, 'fftpack_lite', None)
pfi = getattr(numpy.fft, '_pocketfft_internal', None)
pfi = getattr(numpy.fft, 'pocketfft_internal', pfi)

def run_benchmark(f, x, n):
	x_backup = +x
	y = None
	t = default_timer()
	for i in range(n):
		y = f(x)
	dur = default_timer() - t
	assert numpy.array_equal(x, x_backup), "array should not have been mutated"
	assert numpy.allclose(numpy.fft.rfft(x), y)
	return dur

def benchmark_forward_fft(shape, iterations):
	x = numpy.random.random(shape)
	m = 1 << int(len(x)).bit_length()
	x = numpy.pad(x, [(0, m - len(x))], mode='constant', constant_values=0.0)
	if fpl is not None:
		t = run_benchmark(lambda x, plan=fpl.rffti(m): fpl.rfftf(x, plan), x, iterations)
		print("%s: fftpack_lite.rfft: %.2f us/iteration" % (numpy.__version__, t * 1.0E6 / iterations,))
	if pfi is not None:
		t = run_benchmark(lambda x: pfi.execute(x, True, True, 1), x, iterations)
		print("%s: pocketfft.execute: %.2f us/iteration" % (numpy.__version__, t * 1.0E6 / iterations,))

def main(program, shape = 100, iterations = 1 << 18):
	return benchmark_forward_fft(int(shape), int(iterations))

if __name__ == '__main__':
	import sys
	raise SystemExit(main(*sys.argv) or 0)
```
Running this, I get results similar to the following:
```
1.16.6: fftpack_lite.rfft: 1.22 us/iteration
1.17.0: pocketfft.execute: 2.21 us/iteration
```
This shows a > 80% drop in performance from the past, which seems unfortunate. However, I noticed that pocketfft [was claimed to be faster than NumPy's existing implementation](/numpy/numpy/pull/11888#issuecomment-419048143), which seemed confusing in light of the performance drop, so I thought I'd report this. Or perhaps this is a known issue?

**Comment by mreineck** [comment]
I do not think that this kind of benchmark is helpful in determining the actual speed of the FFT implementation. Calling into a compiled function from Python for a single FFT of length 100 means that most of the time will not be spent computing the transform but rather checking function arguments, preparing data for the call into C etc. 

To demonstrate this, consider:
```
a=np.zeros((1000,100))
%timeit  np.fft.rfft(a,axis=1)
```
This performs 1000 FFTs of length 100 in one go, and on my laptop it takes around 243 microseconds (with pocketfft, I don't have a `numpy` with fftpack installed at the moment). In other words, the real computation time for a single length 100 FFT is a quarter of a microsecond, several times faster than your measured results.

I can imagine two reasons why the `fftpack` version is faster:
- less glue code on the Python side (this has changed a bit in the transition to `pocketfft`). It may be possible to tweak this further, but I'm not sure it's worth the effort
- slower plan generation code in `pocketfft` vs. `fftpack`. This mainly involves computing "twiddle factors", i.e. calling a lot of trigonometric functions. `pocketfft` is emphasizing accuracy here and will probably be slower, while `fftpack` (at least the real-valued FFTs, which you are using) has a pretty bad reputation for accuracy..

I'd argue that if a user requires high performance on small 1D FFTs they should do many of them simultaneously if at all possible. I know that this cannot always be done, but the remaining situations are (in my opinion, and I'm open to counterexamples!) so rare that a general purpose code like `numpy` doesn't need to be specially tweaked for them.

**Comment by mehrdadn** [comment]
Thank you for the quick reply! To be clear, I saw a performance slowdown in a project, and then diagnosed the issue to be pocketfft, and then wrote a simplified benchmark to illustrate it—not the other way around. It was very much the "actual speed" of the code that was slow, and the I spent several hours reducing it to this benchmark so I could demonstrate it in a few lines. The actual project was around ~2000 lines of code, and I felt if I pasted that, I would be asked to trim it down.

If you'd like to see the actual project where this came up, the code I reduced it from is [here](/mehrdadn/SOTA-Py) which you may be able to try out for yourself. The order in which I previously perform fallbacks for the FFT implementations I use is SciPy fftpack → NumPy fftpack_lite → NumPy fft (as the last one is the slowest). As a brief summary, this program easily performs _tens or hundreds of thousands_ of FFTs to compute optimal paths in a stochastic graph. The graph is a road network, and each road requires a convolution of the travel time distribution and the arrival time vectors. These are not simple 1-shot convolutions as they have inter-dependencies—the details are difficult to summarize here, but you can find them in the linked paper & the nested references. The elements in these vectors have cross-dependencies on neighboring vectors, and as such, FFTs need to be performed in a particular manner—it is, in a sense, a manifestation of stochastic dynamic programming (DP).

If you run the program in a virtualenv, one with NumPy based on pocketfft and one with fftpack, and execute it via the given command line, you should be able to see the running time change quite noticeably. (Do not need to install any other packages like SciPy or anything else.)

**Comment by mreineck** [comment]
Thanks for the explanation! I'll fdefinitely have a look at the code.

Meanwhile I have run your benchmark locally with both old and new `numpy` versions, and see:
```
1.16.6: fftpack_lite.rfft: 1.28 us/iteration
1.22.0: pocketfft.execute: 1.25 us/iteration
```
which looks OK. My CPU is an AMD Ryzen 7 4800H ... on which CPU are you running?

**Comment by mehrdadn** [comment]
That's interesting! I wonder why that is. I'm on an Intel i7-7820HQ, on Windows 10 21H2 (19044.1415).

**Comment by mreineck** [comment]
> As a brief summary, this program easily performs tens or hundreds of thousands of FFTs to compute optimal paths in a stochastic graph. The graph is a road network, and each road requires a convolution of the travel time distribution and the arrival time vectors. These are not simple 1-shot convolutions as they have inter-dependencies

This sounds like one of the situations where FFTW's approach with explicit plan generation and management should really shine... but I'm not totally sure, since at very small FFT lengths the calling overhead really becomes the dominant issue and I don't know how much `pyfftw` has been tweaked in this regard.

**Comment by mreineck** [comment]
> That's interesting! I wonder why that is. I'm on an Intel i7-7820HQ, on Windows 10 21H2 (19044.1415).

Just on the off chance ... is this built as a 32bit application? In that case the compiler cannot use SSE2 instructions (unless it is explicitly allowed to), and this could really slow things down.

**Comment by mehrdadn** [comment]
It's possible FFTW would perform better. In fact I may have already tried this at some point; I originally wrote this as a C++ program, and I don't remember everything I tried at the moment, but I definitely recall looking into FFTW at some point. However, it's not necessary for better performance—fftpack already achieved better performance than pocketfft.

Regarding stochastic dynamic programming (SDP), you might be interested in reading [this paper on zero-delay convolution](https://people.cs.clemson.edu/~bcdean/sdp.pdf). While this particular algorithm is not strictly necessary for solving SDP problems, it illustrates the dependencies they inherently have. (You can solve them with both regular and zero-delay convolution; the latter has better asymptotic running time, but the former is faster for smaller vectors.)

And no, everything is 64-bit.

**Comment by mreineck** [comment]
In that case my only idea is some fancy misoptimization in pocketfft by the (quite old) MSVC that's used to build the `numpy` wheels. That could be cross-checked by trying your benchmark under Linux on your CPU. Unfortunately I don't have anything comparable at hand right now.

**Comment by mehrdadn** [comment]
You might be right be right about this! On WSL with 1.21.2 I get 1.26 us/iteration. I wonder if pocketfft is poorly optimized in the Windows build? It might be a while before I get the chance but I will try to get back to you. (I also wonder if we could use GitHub CI to run benchmarks and compare... if I get the chance I might try that.) Thanks!

**Comment by matthew-brett** [comment]
It might not be too hard to use a mingw-w64 UCRT toolchain to compile Numpy on Windows - we're doing that for Scipy, in the recently-merged meson build branch.

**Comment by mreineck** [comment]
> I wonder if pocketfft is poorly optimized in the Windows build? 

It's standard C99 without system calls (except for new/delete, I assume), and shouldn't care at all about the OS running underneath.

**Comment by mehrdadn** [comment]
I'm not sure I'll get a chance to compile with MinGW, though if I do I'll let you know. That said though, I just tried on Ubuntu 20.04 WSL with `python3-scipy` (previously I was trying Arch) and got this:
```
$ python3 Temp.py
1.17.4: pocketfft.execute: 3.79 us/iteration
```
For comparison, previously I was on Arch:
```
$ python3 Temp.py
1.21.2: pocketfft.execute: 1.26 us/iteration
```

**Comment by mreineck** [comment]
In `scipy.fft` I expect somewhat slower execution for short 1D transforms; the calling overhead is higher than in `numpy.fft` because there is much more stuff to be taken care of (support for different data types, potential multithreading etc.)

But the version numbers seem to indicate that you are testing `numpy.fft`, not `scipy.fft`, right?

On Arch, everything is compiled directly for the target CPU, correct? If so, that may give you quite a performance boost.

**Comment by mreineck** [comment]
Hypothesis: if the performance discrepancies only exist for short FFTs and go away for, say, length 1000 and 10000, I suspect that the culprit is the function `my_sincosm1pi()`. This is pretty low-level code and different compilers will probably compile this very differently (also depending on compiler flags).

**Comment by mehrdadn** [comment]
Yes I'm running `numpy.fft` and not `scipy.fft` in that benchmark. Arch is a binary rolling-release distro similar to Debian-unstable, so it's not hyper-optimized for a particular CPU to my knowledge. (It's not source-based like Gentoo.)

The performance discrepancy indeed does appear to diminish as the length increases, so you may be correct about that! Unfortunately to confirm this I'll probably have to perform at least 2 builds with debug symbols that exhibit the discrepancy, and it might well be quite a while before I get a chance to do that. I'll be happy to get back to you though if I manage to. Thank you for the help so far! It definitely surprised me to see such a difference across platforms.

**Comment by mehrdadn** [comment]
Something interesting I just noticed in `my_sincosm1pi` is that it calls `fma`, and my `pocketfft_internal.cp36-win_amd64.pyd` DLL imports `fma` as an external function from `api-ms-win-crt-math-l1-1-0.dll`. I don't yet know if it's supposed to be inlined, but if so, that might well explain it.

**Comment by mreineck** [comment]
Perhaps it wasn't such a great idea to use hyper-accurate, but also hairy-to-compile `my_sincosm1pi`; there is not much loss of accuracy when switching to plain `sin/cos`, and compilers might have a better chance to optimize those. If `fma` is not inlined, that is most likely the cause for the slowdown. On the other hand, I cannot imagine *any* compiler that doesn't inline `sqrt`, so I'm not totally sure what these external symbols mean.

**Comment by mehrdadn** [comment]
Under `/fp:fast`, Visual C++ seems to inline `sqrt`. But under `/fp:precise`, which is often the default, it only performs a partial inining (I just compiled a test program and saw this). It generates the native instruction for it too, but it compares it against 0 and calls the function in one branch. I haven't spent the time to figure out why yet, but I know [this page](https://docs.microsoft.com/en-us/cpp/build/reference/fp-specify-floating-point-behavior) has some info that might be relevant. The instructions in question end up being:
```
00007FF7496FA5D4 66 0F 6E C9          movd        xmm1,ecx  
00007FF7496FA5D8 F3 0F E6 C9          cvtdq2pd    xmm1,xmm1  
00007FF7496FA5DC F2 0F 5E 0D 34 A8 00 00 divsd       xmm1,mmword ptr [__real@4059000000000000 (07FF749704E18h)]  
00007FF7496FA5E4 0F 57 C0             xorps       xmm0,xmm0  
00007FF7496FA5E7 66 0F 2E C1          ucomisd     xmm0,xmm1  
00007FF7496FA5EB 77 0D                ja          wmain+4Ah (07FF7496FA5FAh)  
00007FF7496FA5ED 66 0F 51 C1          sqrtpd      xmm0,xmm1  
00007FF7496FA5F1 F2 0F 2C C0          cvttsd2si   eax,xmm0  
00007FF7496FA5F5 E9 82 03 00 00       jmp         wmain+3CCh (07FF7496FA97Ch)  
00007FF7496FA5FA 0F 28 C1             movaps      xmm0,xmm1  
00007FF7496FA5FD E8 01 A8 FC FF       call        sqrt (07FF7496C4E03h)  
00007FF7496FA602 F2 0F 2C C0          cvttsd2si   eax,xmm0  
00007FF7496FA606 E9 71 03 00 00       jmp         wmain+3CCh (07FF7496FA97Ch)  
```

**Comment by mehrdadn** [comment]
That said, this may or may not be the issue; I'd have to build with symbols to check.

**Comment by mehrdadn** [comment]
I tried compiling `_pocketfft.c` with symbols (didn't recompile other parts of NumPy) and unfortunately I didn't manage to track it down.

I found the effects of all of the following were negligible:

- Reimplementing `my_sincosm1pi` via `sin` and `cos`
- Reimplementing `my_sincosm1pi` via `fsincos`
- Reimplementing `fma` via `_mm_fmadd_pd`
- Compiling with `-fp:fast`
- Compiling with `-arch:AVX2`
- Compiling with `-GL`
- Compiling with `clang-cl`

Furthermore, the behavior on Ubuntu 22.04 (`python3-numpy 1:1.21.5-1build2`) matches that on Arch (`python-numpy 1.21.2-1`), and the performance of Windows (`CPython v3.10.0:b494f59` on [NumPy `1.21.5+mkl`](https://www.lfd.uci.edu/~gohlke/pythonlibs/#cython)) is consistently slower than both of them.

Unfortunately I probably won't be digging into this anymore, but in case anyone would like to investigate, here are _possible_ hotspots I saw in the debugger:
- Various functions in `_multiarray_umath`
- `malloc`/`free` (both called from `_multiarray_umath` and also from `_pocketfft`)
This should be much easier to track down if someone is able to compile all of NumPy with symbols on both Windows and Linux, and then use VTune on both platforms to compare the hotspots.

**Comment by mehrdadn** [comment]
Yeah, SciPy's version seems to be slower across the board:
```
Windows
[SciPy 1.7.3 , Python 3.10.0]   scipy.fft._pocketfft.basic.r2c  33.6 us/iteration
[NumPy 1.21.5, Python 3.10.0]              pocketfft.rfftf      25.9 us/iteration
[NumPy 1.21.5, Python 3.10.0]              numpy.fft.rfft       26.2 us/iteration

Arch
[SciPy 1.7.2 , Python 3.9.7]    scipy.fft._pocketfft.basic.r2c  25.4 us/iteration
[NumPy 1.21.2, Python 3.9.7]               pocketfft.rfftf      17.6 us/iteration
[NumPy 1.21.2, Python 3.9.7]               numpy.fft.rfft       21.7 us/iteration

Ubuntu
[SciPy 1.8.0 , Python 3.10.4]   scipy.fft._pocketfft.basic.r2c  24.0 us/iteration
[NumPy 1.21.5, Python 3.10.4]              pocketfft.rfftf      18.2 us/iteration
[NumPy 1.21.5, Python 3.10.4]              numpy.fft.rfft       21.4 us/iteration
```

**Comment by seberg** [comment]
I think your FFTs are pretty small?  This would seem like changes in the overhead.  The big chunk is likely the `__array_function__` overhead ([which could be improved](gh-20790)).
You could try to run with `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=0` as an environment variable and see if it makes a difference in speed.  That would show whether it is (largely) due to the array function kicking in.

**Comment by mehrdadn** [comment]
Yes, I think I've basically ruled out the arithmetic as the culprit through the changes I tested and mentioned. That's why I pointed to heap allocations and `_multiarray_umath` above.

`NUMPY_EXPERIMENTAL_ARRAY_FUNCTION` doesn't appear to make any difference. The difference between Windows & Linux does go down as the FFT size gets larger, but it has to get *way* larger for this to go down substantially. For convolving arrays of sizes 100 and 70 it's a ~25% difference. For 1000 and 700 it goes down to 20%. For 10000 and 7000 it goes down to < 3% (sometimes 0% within the variance).

SciPy does seem to become considerably faster (over 2x) for larger arrays.

I'd like to note that "small" arrays (~100 elements) are valid use cases to worry about, though. The fact that they aren't several thousand+ elements each doesn't mean their performance is irrelevant.

**Comment by seberg** [comment]
Ah, so this was the precision discussion above, that calculating some factors ended up in some very slow paths in windows.

**Comment by mattip** [comment]
This issue has wandered around various OSs, python versions, and FFT implementations and is quite old. I will close it. Please open a new issue if some slowdown can be clearly identified.
